app.js

```diff
@@ -179,44 +180,63 @@ app.delete('/api/v1/projects/:id', (request, response) => {
 });
 
 app.post('/api/v1/users', (request, response) => {
-   const user = request.body;
-
-  for (let param of ['username', 'password']) {
-    if (!user[param]) {
-      return response.status(422).json({ error: `Expected body to = { username: <string>, password: <string> }. Body is missing a '${param}' param.` });
-    } 
-  }
-
-  database('users').where({ username: user.username }).first()
-    .then(user => {
-      if (user) {
-        return response.status(409).json({ error: `Resource already exists with username of '${user.username}'.` });
-      }
-    });
+  const { username, password } = request.body;
+
+  bcrypt.hash(password, 10, (err, hash) => {
+    const user = {
+      username,
+      password: hash
+    };
+
+    for (let param of ['username', 'password']) {
+      if (!user[param]) {
+        return response.status(422).json({ error: `Expected body to = { username: <string>, password: <string> }. Body is missing a '${param}' param.` });
+      } 
+    }
 
-  database('users').insert(user, 'id')
-    .then(user => {
-      response.status(201).json({ id: user[0] });
-    })
-    .catch(error => {
-      response.status(500).json({ error });
-    });
+    database('users').where({ username: user.username }).first()
+      .then(user => {
+        if (user) {
+          return response.status(409).json({ error: `Resource already exists with username of '${user.username}'.` });
+        }
+      });
+
+    database('users').insert(user, 'id')
+      .then(user => {
+        response.status(201).json({ id: user[0] });
+      })
+      .catch(error => {
+        response.status(500).json({ error });
+      });
+   });
 });
 
-app.get('/api/v1/users/:username', (request, response) => {
-  const username = request.params.username;
+app.get('/api/v1/users/login', (request, response) => {
+  const { username, password } = request.body;
 
   database('users').where({ username }).first()
     .then(user => {
-      if (user) {
-        response.status(200)
+      if (!user) {
+        return response.status(404)
```

Comment:
Really smart to give a status of 404 for username not matching before checking the password